src/apps/lwaftr/fragmentv6_hardened.lua

```diff
@@ -0,0 +1,299 @@
+   local o_id = ehs + ipv6_fixed_header_size + o_ipv6_frag_id
+   ffi.copy(key.src_addr, fragment.data + o_src, 16)
+   ffi.copy(key.dst_addr, fragment.data + o_dst, 16)
+   key.fragment_id = C.ntohl(rd32(fragment.data + o_id))
+   return key
+end
+
+local function free_reassembly_buf_and_pkt(pkt, frags_table)
+   local key = get_key(pkt)
+   frags_table:remove(key, false)
+   packet.free(pkt)
+end
+
+local function swap(array, i, j)
+   local tmp = array[j]
+   array[j] = array[i]
+   array[i] = tmp
+end
+
+-- This is an insertion sort, and only called on 2+ element arrays
+local function sort_array(array, last_index)
+   for i=0,last_index do
+      local j = i
+      while j > 0 and array[j-1] > array[j] do
+         swap(array, j, j-1)
+         j = j - 1
+      end
+   end
+end
+
+local function verify_valid_offsets(reassembly_buf)
+   if reassembly_buf.fragment_starts[0] ~= 0 then
+      return false
+   end
+   for i=1,reassembly_buf.fragment_count-1 do
+      if reassembly_buf.fragment_starts[i] ~= reassembly_buf.fragment_ends[i-1] then
+         return false
+      end
+   end
+   return true
+end
+
+local function reassembly_status(reassembly_buf)
+   if reassembly_buf.final_start == 0 then
+      return FRAGMENT_MISSING
+   end
+   if reassembly_buf.running_length ~= reassembly_buf.reassembly_packet.length then
+      return FRAGMENT_MISSING
+   end
+   if not verify_valid_offsets(reassembly_buf) then
+      return REASSEMBLY_INVALID
+   end
+   return REASSEMBLY_OK
+end
+
+local function attempt_reassembly(frags_table, reassembly_buf, fragment)
+   local reassembly_pkt = reassembly_buf.reassembly_packet
+   local frag_id = get_frag_id(fragment)
+   if frag_id ~= reassembly_buf.fragment_id then -- unreachable
+      assert(false, "Impossible case reached in v6 reassembly")
+      return REASSEMBLY_INVALID
+   end
+
+   local frag_start = get_frag_start(fragment)
+   local frag_size = get_frag_len(fragment)
+   local fcount = reassembly_buf.fragment_count
+   if fcount + 1 > max_frags_per_packet then
+      -- too many fragments to reassembly this packet, assume malice
+      free_reassembly_buf_and_pkt(fragment, frags_table)
+      return REASSEMBLY_INVALID
+   end
+   reassembly_buf.fragment_starts[fcount] = frag_start
+   reassembly_buf.fragment_ends[fcount] = frag_start + frag_size
+   if reassembly_buf.fragment_starts[fcount] <
+      reassembly_buf.fragment_starts[fcount - 1] then
+      sort_array(reassembly_buf.fragment_starts, fcount)
+      sort_array(reassembly_buf.fragment_ends, fcount)
+   end
+   reassembly_buf.fragment_count = fcount + 1
+   if is_last_fragment(fragment) then
+      if reassembly_buf.final_start ~= 0 then
+         -- There cannot be 2+ final fragments
+         free_reassembly_buf_and_pkt(fragment, frags_table)
+         return REASSEMBLY_INVALID
+      else
+         reassembly_buf.final_start = frag_start
+      end
+   end
+
+   -- This is a massive layering violation. :/
```

Comment:
Maybe some explanation about why that's the case, and why it's still the lesser evil?